README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-blue?style=flat-square)](LICENSE)
[![Crates.io](https://img.shields.io/crates/v/tasd?style=flat-square)](https://crates.io/crates/tasd)
[![Documentation](https://img.shields.io/docsrs/tasd?style=flat-square)](https://docs.rs/tasd)
### Description
Encoder/decoder crate for the [TASD](https://github.com/bigbass1997/TASD-Spec) file format specification.

Highest format version supported: **0x0001**

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## [0.4.0] - 2023-10-13
- Added strum EnumString to PacketKind to allow converting to/from a string
- Added method for writing a string prefixed with a u8 length byte
- Updated GAME_IDENTIFIER to include a name
- Changed lookup functions to return an owned value

## [0.3.0] - 2023-09-28
- Added unit and integration tests (in-progress)
- Fixed bug where payload lengths larger than 255 bytes were decoded incorrectly
- Removed NES_OVERREAD and SNES_OVERREAD
- Added PORT_OVERREAD
- Changed TRANSITION to include a port number

## [0.2.1] - 2023-07-12
- Changed `TasdFile::new()` to include the DUMP_CREATED packet in the returned `TasdFile` instance
- Changed PacketKind's Display implementation to return the name in SNAKE_CASE
- Added `From<T> for Packet` implementations for all packet structs

## [0.2.0] - 2023-07-06
- Changed packet data structures
- Added SNES_LATCH_TRAIN packet


tests/packets.rs
use tasd::spec::packets::{Attribution, Category, ConsoleRegion, ConsoleType, Encode, GameTitle, Packet, RomName};
use tasd::spec::writer::Writer;

/// Small wrapper around [`Writer`] for creating a packet using a key and some data.
fn packet<D: AsRef<[u8]>>(key: &[u8], data: D) -> Vec<u8> {
    let mut w = Writer::new();
    w.write_slice(data.as_ref());
    
    w.into_packet(key, key.len() as u8)
}

macro_rules! assert_packet {
    ($packet:expr, $key:expr, $data:expr) => {
        assert_eq!(
            Packet::from($packet).encode($key.len() as u8),
            packet(&$key, $data),
        );
    };
}

#[test]
fn console_type() {
    assert_packet!(ConsoleType { kind: 0x01, custom: None }, [0x00, 0x01], [0x01]);
    assert_packet!(ConsoleType { kind: 0x05, custom: None }, [0x00, 0x01], [0x05]);
    assert_packet!(ConsoleType { kind: 0xFF, custom: None }, [0x00, 0x01], [0xFF]);
    assert_packet!(ConsoleType { kind: 0xFF, custom: Some("something".into()) }, [0x00, 0x01], [&[0xFF], "something".as_bytes()].concat());
}

#[test]
fn console_region() {
    assert_packet!(ConsoleRegion { region: 0x01 }, [0x00, 0x02], [0x01]);
    assert_packet!(ConsoleRegion { region: 0x02 }, [0x00, 0x02], [0x02]);
    assert_packet!(ConsoleRegion { region: 0xFF }, [0x00, 0x02], [0xFF]);
}

#[test]
fn game_title() {
    assert_packet!(GameTitle { title: "This is the title!".into() }, [0x00, 0x03], "This is the title!".as_bytes());
    assert_packet!(GameTitle { title: "12345? That's amazing, I've got the same password on my luggage!".into() }, [0x00, 0x03], "12345? That's amazing, I've got the same password on my luggage!".as_bytes());
}

#[test]
fn rom_name() {
    assert_packet!(RomName { name: "Super Mario Bros. Super Hack World With Ultra Mods".into() }, [0x00, 0x04], "Super Mario Bros. Super Hack World With Ultra Mods".as_bytes());
}

#[test]
fn attribution() {
    assert_packet!(Attribution { kind: 0x01, name: "Arthur".into() }, [0x00, 0x05], [&[0x01], "Arthur".as_bytes()].concat());
    assert_packet!(Attribution { kind: 0x02, name: "a replay device".into() }, [0x00, 0x05], [&[0x02], "a replay device".as_bytes()].concat());
    assert_packet!(Attribution { kind: 0x03, name: "Arthur".into() }, [0x00, 0x05], [&[0x03], "Arthur".as_bytes()].concat());
    assert_packet!(Attribution { kind: 0x04, name: "Arthur".into() }, [0x00, 0x05], [&[0x04], "Arthur".as_bytes()].concat());
    assert_packet!(Attribution { kind: 0xFF, name: "Arthur".into() }, [0x00, 0x05], [&[0xFF], "Arthur".as_bytes()].concat());
}

#[test]
fn category() {
    assert_packet!(Category { category: "any%, no jumps, invisible hud, 2 players".into() }, [0x00, 0x06], "any%, no jumps, invisible hud, 2 players".as_bytes());
}

#[test]
fn emulator_name() {
    
}

#[test]
fn emulator_version() {
    
}

#[test]
fn emulator_core() {
    
}

#[test]
fn tas_last_modified() {
    
}

#[test]
fn dump_created() {
    
}

#[test]
fn dump_last_modified() {
    
}

#[test]
fn total_frames() {
    
}

#[test]
fn rerecords() {
    
}

#[test]
fn source_link() {
    
}

#[test]
fn blank_frames() {
    
}

#[test]
fn verified() {
    
}

#[test]
fn memory_init() {
    
}

#[test]
fn game_identifier() {
    
}

#[test]
fn movie_license() {
    
}

#[test]
fn movie_file() {
    
}

#[test]
fn port_controller() {
    
}

#[test]
fn nes_latch_filter() {
    
}

#[test]
fn nes_clock_filter() {
    
}

#[test]
fn nes_overread() {
    
}

#[test]
fn nes_game_genie_code() {
    
}

#[test]
fn snes_clock_filter() {
    
}

#[test]
fn snes_overread() {
    
}

#[test]
fn snes_game_genie_code() {
    
}

#[test]
fn snes_latch_train() {
    
}

#[test]
fn genesis_game_genie_code() {
    
}

#[test]
fn input_chunk() {
    
}

#[test]
fn input_moment() {
    
}

#[test]
fn transition() {
    
}

#[test]
fn lag_frame_chunk() {
    
}

#[test]
fn movie_transition() {
    
}

#[test]
fn comment() {
    
}

#[test]
fn experimental() {
    
}

#[test]
fn unspecified() {
    
}

#[test]
fn unsupported() {
    
}


src/spec/reader.rs


pub struct Reader<'a> {
    inner: &'a [u8],
    pos: usize,
}
impl<'a> Reader<'a> {
    pub fn new<T: AsRef<[u8]>>(inner: &'a T) -> Self {
        Self {
            inner: inner.as_ref(),
            pos: 0,
        }
    }
    
    pub fn peek_u8(&self) -> u8 {
        self.inner[self.pos]
    }
    
    pub fn peek_u16(&self) -> u16 {
        u16::from_be_bytes(self.inner[self.pos..(self.pos + 2)].try_into().unwrap())
    }
    
    pub fn peek_u32(&self) -> u32 {
        u32::from_be_bytes(self.inner[self.pos..(self.pos + 4)].try_into().unwrap())
    }
    
    pub fn peek_u64(&self) -> u64 {
        u64::from_be_bytes(self.inner[self.pos..(self.pos + 8)].try_into().unwrap())
    }
    
    pub fn peek_i8(&self) -> i8 {
        self.peek_u8() as i8
    }
    
    pub fn peek_i16(&self) -> i16 {
        self.peek_u16() as i16
    }
    
    pub fn peek_i32(&self) -> i32 {
        self.peek_u32() as i32
    }
    
    pub fn peek_i64(&self) -> i64 {
        self.peek_u64() as i64
    }
    
    /// Peeks `len` bytes starting from the current position.
    pub fn peek_len(&self, len: usize) -> &[u8] {
        &self.inner[self.pos..(self.pos + len)]
    }
    
    /// Peeks `len` bytes starting from the current position, and returns the bytes in reversed order.
    /// 
    /// # Example
    /// ```
    /// use tasd::spec::reader::Reader;
    /// 
    /// let mut r = Reader::new(&[0x11, 0x22, 0x33, 0x44, 0x55]);
    /// r.advance(1);
    /// assert_eq!(r.peek_len(3),     &[0x22, 0x33, 0x44]);
    /// assert_eq!(r.peek_len_rev(3), &[0x44, 0x33, 0x22]);
    /// ```
    pub fn peek_len_rev(&self, len: usize) -> Vec<u8> {
        self.peek_len(len).iter().copied().rev().collect()
    }
    
    
    pub fn read_u8(&mut self) -> u8 {
        let data = self.inner[self.pos];
        self.pos += 1;
        
        data
    }
    
    pub fn read_u16(&mut self) -> u16 {
        let data = u16::from_be_bytes(self.inner[self.pos..(self.pos + 2)].try_into().unwrap());
        self.pos += 2;
        
        data
    }
    
    pub fn read_u32(&mut self) -> u32 {
        let data = u32::from_be_bytes(self.inner[self.pos..(self.pos + 4)].try_into().unwrap());
        self.pos += 4;
        
        data
    }
    
    pub fn read_u64(&mut self) -> u64 {
        let data = u64::from_be_bytes(self.inner[self.pos..(self.pos + 8)].try_into().unwrap());
        self.pos += 8;
        
        data
    }
    
    pub fn read_i8(&mut self) -> i8 {
        self.read_u8() as i8
    }
    
    pub fn read_i16(&mut self) -> i16 {
        self.read_u16() as i16
    }
    
    pub fn read_i32(&mut self) -> i32 {
        self.read_u32() as i32
    }
    
    pub fn read_i64(&mut self) -> i64 {
        self.read_u64() as i64
    }
    
    
    pub fn read_bool(&mut self) -> bool {
        self.read_u8() > 0
    }
    
    pub fn read_len(&mut self, len: usize) -> &[u8] {
        let data = &self.inner[self.pos..(self.pos + len)];
        self.pos += len;
        
        data
    }
    
    pub fn read_string(&mut self, len: usize) -> String {
        String::from_utf8_lossy(self.read_len(len)).to_string()
    }
    
    pub fn read_remaining(&mut self) -> &[u8] {
        let data = &self.inner[self.pos..];
        self.pos += self.remaining();
        
        data
    }
    
    /// Copies entire buffer into a Vec regardless of current position.
    /// 
    /// Use [`Self::read_remaining`] if only the remaining data is needed.
    pub fn to_vec(&self) -> Vec<u8> {
        self.inner.to_vec()
    }
    
    pub fn advance(&mut self, count: usize) {
        self.pos += count;
    }
    
    pub fn rewind(&mut self, count: usize) {
        self.pos -= count;
    }
    
    pub fn remaining(&self) -> usize {
        self.inner.len() - self.pos
    }
    
    pub fn pos(&self) -> usize {
        self.pos
    }
    
    pub fn set_pos(&mut self, pos: usize) {
        self.pos = pos;
    }
}





#[cfg(test)]
mod tests {
    use crate::spec::reader::Reader;
    
    const TEST_DATA: [[u8; 16]; 3] = [
        [0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF],
        [0xF0, 0xE1, 0xD2, 0xC3, 0xB4, 0xA5, 0x96, 0x87, 0x78, 0x69, 0x5A, 0x4B, 0x3C, 0x2D, 0x1E, 0x0F],
        [0x00, 0x01, 0x01, 0x00, 0x00, 0x01, 0xFF, 0x00, 0xAA, 0x55, 0x00, 0x69, 0x96, 0x00, 0x01, 0x00],
    ];
    
    #[test]
    fn controls() {
        for data in TEST_DATA {
            let mut r = Reader::new(&data);
            
            assert_eq!(r.pos, 0);
            assert_eq!(r.pos, r.pos());
            assert_eq!(r.inner, &data);
            
            let mut i = 0;
            for adv in [0, 0, 1, 1, 5, 23, 1963, 9] {
                r.advance(adv);
                i += adv;
                assert_eq!(r.pos, i);
                assert_eq!(r.pos, r.pos());
            }
            
            for set in [592, 11, 50, 99, 2, 0, 8, 9102502] {
                r.set_pos(set);
                assert_eq!(r.pos, set);
                assert_eq!(r.pos, r.pos());
            }
            
            r.set_pos(usize::MAX);
            let mut i = r.pos();
            for rev in [0, 1, 53, 3, 0] {
                r.rewind(rev);
                i -= rev;
                assert_eq!(r.pos, i);
                assert_eq!(r.pos, r.pos());
            }
            
            r.set_pos(0);
            for i in 0..data.len() {
                assert_eq!(r.remaining(), data.len() - i);
                r.advance(1);
            }
            
            r.set_pos(data.len());
            for i in 0..data.len() {
                assert_eq!(r.remaining(), i);
                r.rewind(1);
            }
        }
    }
    
    #[test]
    fn peeks() {
        for data in TEST_DATA {
            let mut r = Reader::new(&data);
            
            r.set_pos(0);
            for i in 0..data.len() {
                assert_eq!(r.peek_u8(), data[i]);
                assert_eq!(r.peek_i8(), data[i] as i8);
                r.advance(1);
            }
            
            r.set_pos(0);
            for i in 0..(data.len() - 1) {
                let expected = u16::from_be_bytes(data[i..(i + 2)].try_into().unwrap());
                assert_eq!(r.peek_u16(), expected);
                assert_eq!(r.peek_i16(), expected as i16);
                r.advance(1);
            }
            
            r.set_pos(0);
            for i in 0..(data.len() - 3) {
                let expected = u32::from_be_bytes(data[i..(i + 4)].try_into().unwrap());
                assert_eq!(r.peek_u32(), expected);
                assert_eq!(r.peek_i32(), expected as i32);
                r.advance(1);
            }
            
            r.set_pos(0);
            for i in 0..(data.len() - 7) {
                let expected = u64::from_be_bytes(data[i..(i + 8)].try_into().unwrap());
                assert_eq!(r.peek_u64(), expected);
                assert_eq!(r.peek_i64(), expected as i64);
                r.advance(1);
            }
            
            r.set_pos(0);
            for i in 0..data.len() {
                assert_eq!(r.peek_len(i), &data[..i]);
                assert_eq!(r.peek_len_rev(i), data[..i].into_iter().copied().rev().collect::<Vec<u8>>());
            }
        }
    }
    
    #[test]
    fn reads() {
        for data in TEST_DATA {
            let mut r = Reader::new(&data);
            
            r.set_pos(0);
            for i in 0..data.len() {
                assert_eq!(r.read_u8(), data[i]);
                r.rewind(1);
                assert_eq!(r.read_i8(), data[i] as i8);
                r.rewind(1);
                assert_eq!(r.read_bool(), data[i] > 0);
            }
            
            r.set_pos(0);
            for i in 0..(data.len() - 1) {
                let expected = u16::from_be_bytes(data[i..(i + 2)].try_into().unwrap());
                assert_eq!(r.read_u16(), expected);
                r.rewind(2);
                
                let expected = i16::from_be_bytes(data[i..(i + 2)].try_into().unwrap());
                assert_eq!(r.read_i16(), expected);
                r.rewind(1);
            }
            
            r.set_pos(0);
            for i in 0..(data.len() - 3) {
                let expected = u32::from_be_bytes(data[i..(i + 4)].try_into().unwrap());
                assert_eq!(r.read_u32(), expected);
                r.rewind(4);
                
                let expected = i32::from_be_bytes(data[i..(i + 4)].try_into().unwrap());
                assert_eq!(r.read_i32(), expected);
                r.rewind(3);
            }
            
            r.set_pos(0);
            for i in 0..(data.len() - 7) {
                let expected = u64::from_be_bytes(data[i..(i + 8)].try_into().unwrap());
                assert_eq!(r.read_u64(), expected);
                r.rewind(8);
                
                let expected = i64::from_be_bytes(data[i..(i + 8)].try_into().unwrap());
                assert_eq!(r.read_i64(), expected);
                r.rewind(7);
            }
            
            for i in 0..data.len() {
                r.set_pos(0);
                assert_eq!(r.read_len(i), &data[..i]);
                
                r.set_pos(data.len() - i);
                assert_eq!(r.read_len(i), &data[(data.len() - i)..]);
                
                r.set_pos(i);
                assert_eq!(r.read_remaining(), &data[i..]);
                
                r.set_pos(0);
                assert_eq!(r.read_string(i), String::from_utf8_lossy(&data[..i]));
            }
        }
    }
    
    #[test]
    fn conversion() {
        for data in TEST_DATA {
            let mut r = Reader::new(&data);
            
            for i in 0..data.len() {
                r.set_pos(i);
                assert_eq!(r.read_remaining(), &data[i..]);
                assert_eq!(r.read_remaining(), &[]);
                
                assert_eq!(r.to_vec(), &data);
            }
        }
    }
}

src/spec/writer.rs
use std::cmp::{max, min};
use crate::util::to_bytes;

pub struct Writer {
    inner: Vec<u8>,
}
impl Writer {
    pub fn new() -> Self {
        Self {
            inner: vec![]
        }
    }
    
    pub fn write_u8(&mut self, data: u8) {
        self.inner.push(data);
    }
    
    pub fn write_u16(&mut self, data: u16) {
        self.inner.extend_from_slice(&data.to_be_bytes());
    }
    
    pub fn write_u32(&mut self, data: u32) {
        self.inner.extend_from_slice(&data.to_be_bytes());
    }
    
    pub fn write_u64(&mut self, data: u64) {
        self.inner.extend_from_slice(&data.to_be_bytes());
    }
    
    pub fn write_i8(&mut self, data: i8) {
        self.write_u8(data as u8);
    }
    
    pub fn write_i16(&mut self, data: i16) {
        self.write_u16(data as u16);
    }
    
    pub fn write_i32(&mut self, data: i32) {
        self.write_u32(data as u32);
    }
    
    pub fn write_i64(&mut self, data: i64) {
        self.write_u64(data as u64);
    }
    
    pub fn write_bool(&mut self, data: bool) {
        self.inner.push(data as u8);
    }
    
    pub fn write_str(&mut self, data: &str) {
        self.inner.extend_from_slice(data.as_bytes());
    }
    
    pub fn write_u8_str(&mut self, data: &str) {
        let data = data.as_bytes();
        let len = min(data.len(), 255);
        
        self.write_u8(len as u8);
        self.inner.extend_from_slice(&data[..len]);
    }
    
    pub fn write_option_string(&mut self, data: &Option<String>) {
        self.inner.extend_from_slice(data.as_ref().unwrap_or(&"".into()).as_bytes());
    }
    
    pub fn write_slice(&mut self, data: &[u8]) {
        self.inner.extend_from_slice(data);
    }
    
    pub fn write_iter<I: IntoIterator<Item = u8>>(&mut self, data: I) {
        self.inner.extend(data);
    }
    
    pub fn into_packet(self, key: &[u8], keylen: u8) -> Vec<u8> {
        let key = {
            let mut resized_key = vec![0u8; max(key.len(), keylen as usize) - key.len()];
            resized_key.extend_from_slice(key);
            resized_key
        };
        
        let exp = {
            let mut tmp = self.inner.len();
            let mut exp = 0u8;
            while tmp > 0 {
                tmp >>= 8;
                exp += 1;
            }
            exp
        };
        let plen = to_bytes(self.inner.len(), exp);
        
        let mut data = Vec::with_capacity(self.inner.len() + key.len() + 1 + exp as usize);
        data.extend_from_slice(&key);
        data.push(exp);
        data.extend_from_slice(&plen);
        data.extend_from_slice(&self.inner);
        
        data
    }
    
    /// Returns a clone of this [Writer]'s internal buffer.
    pub fn to_vec(&self) -> Vec<u8> {
        self.inner.clone()
    }
}





#[cfg(test)]
mod tests {
    use std::array::from_fn;
    use std::cmp::min;
    use crate::spec::writer::Writer;
    
    #[test]
    fn writes() {
        fn perform<T: Copy, I: IntoIterator<Item = T>, F: Fn(&mut Writer, &mut Vec<u8>, T)>(pattern: I, func: F) {
            let mut w = Writer::new();
            let mut data = Vec::with_capacity(256);
            for i in pattern.into_iter() {
                func(&mut w, &mut data, i);
            }
            assert_eq!(w.inner, data);
        }
        
        perform(u8::MIN..=u8::MAX, |w, expected, data| {
            w.write_u8(data);
            expected.push(data);
        });
        
        perform(-128..=127, |w, expected, data| {
            w.write_i8(data);
            expected.push(data as u8);
        });
        
        const U16PAT: [u16; 8] = [0x0000, 0x55AA, 0xAA55, 0xFFFF, 0x1234, 0x6969, 0x0FA1, 0x1010];
        perform(U16PAT, |w, expected, data| {
            w.write_u16(data);
            expected.extend_from_slice(&data.to_be_bytes());
        });
        
        const I16PAT: [i16; 8] = [0x0000, 0x55AA, -0x7A55, -0x8000, 0x1234, 0x6969, 0x0FA1, 0x1010];
        perform(I16PAT, |w, expected, data| {
            w.write_i16(data);
            expected.extend_from_slice(&data.to_be_bytes());
        });
        
        const U32PAT: [u32; 8] = [0x00000000, 0x55AA33DD, 0xAA55DD33, 0xFFFFFFFF, 0x12345678, 0x69696969, 0x0FA17C15, 0x10101010];
        perform(U32PAT, |w, expected, data| {
            w.write_u32(data);
            expected.extend_from_slice(&data.to_be_bytes());
        });
        
        const I32PAT: [i32; 8] = [0x00000000, 0x55AA33DD, -0x7A55DD33, -0x80000000, 0x12345678, 0x69696969, 0x0FA17C15, 0x10101010];
        perform(I32PAT, |w, expected, data| {
            w.write_i32(data);
            expected.extend_from_slice(&data.to_be_bytes());
        });
        
        const U64PAT: [u64; 8] = [0x0000000000000000, 0x55AA55AA55AA55AA, 0xAA55AA55AA55AA55, 0xFFFFFFFFFFFFFFFF, 0x123456789ABCDEF, 0x6969696969696969, 0x0FA17C15A6B90D38, 0x1010101010101010];
        perform(U64PAT, |w, expected, data| {
            w.write_u64(data);
            expected.extend_from_slice(&data.to_be_bytes());
        });
        
        const I64PAT: [i64; 8] = [0x0000000000000000, 0x55AA55AA55AA55AA, -0x7A55AA55AA55AA55, -0x8000000000000000, 0x123456789ABCDEF, 0x6969696969696969, 0x0FA17C15A6B90D38, 0x1010101010101010];
        perform(I64PAT, |w, expected, data| {
            w.write_i64(data);
            expected.extend_from_slice(&data.to_be_bytes());
        });
        
        perform([false, true, false, false, true, true, false, false, false, true, true, true], |w, expected, data| {
            w.write_bool(data);
            expected.push(data as u8);
        });
        
        perform(["foo", "bar", "", "fish", "tasd", "hello world!", "lorem ipsum"], |w, expected, data| {
            w.write_str(data);
            expected.extend_from_slice(data.as_bytes());
        });
        
        perform(["foo", "bar", "", "fish", "tasd", "hello world!", "lorem ipsum"], |w, expected, data| {
            w.write_u8_str(data);
            expected.push(data.len() as u8);
            expected.extend_from_slice(data.as_bytes());
        });
        for len in 0..280 {
            let mut w = Writer::new();
            let s = String::from_utf8(vec![0x5A; len]).unwrap();
            w.write_u8_str(&s);
            assert_eq!(w.inner.len(), 1 + min(len, 255));
            
            let mut expected = Vec::with_capacity(512);
            expected.push(min(len, 255) as u8);
            expected.extend_from_slice(&s.as_bytes()[..min(len, 255)]);
            assert_eq!(w.inner, expected);
        }
        
        let mut w = Writer::new();
        w.write_iter(0..=255);
        assert_eq!(w.inner, (0..=255u8).into_iter().collect::<Vec<u8>>());
    }
    
    #[test]
    fn conversion() {
        let data = [0x11, 0x22, 0x33, 0xA5, 0x5A, 0x00, 0xFF];
        let mut w = Writer::new();
        
        w.write_slice(&data);
        assert_eq!(w.to_vec(), data);
        
        let mut w = Writer::new();
        let data: [u8; 0x105A5] = from_fn(|i| i as u8);
        w.write_iter(data.clone());
        
        let mut packet = vec![
            0x5A, 0xA5,
            0x03, 0x01, 0x05, 0xA5
        ];
        packet.extend_from_slice(&data);
        assert_eq!(w.into_packet(&[0x5A, 0xA5], 2), packet);
    }
}

src/spec.rs
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};
use crate::spec::packets::{DumpCreated, Encode, Packet, PacketError};
use crate::spec::reader::Reader;
use crate::spec::writer::Writer;

pub mod packets;
pub mod reader;
pub mod writer;

pub const LATEST_VERSION: [u8; 2] = [0x00, 0x01];
pub const MAGIC_NUMBER: [u8; 4] = [0x54, 0x41, 0x53, 0x44];

#[derive(Debug)]
pub enum TasdError {
    Io(std::io::Error),
    Packet(PacketError),
    MissingHeader,
    MagicNumberMismatch(Vec<u8>),
    MissingPath,
}
impl From<std::io::Error> for TasdError {
    fn from(value: std::io::Error) -> Self {
        Self::Io(value)
    }
}
impl From<PacketError> for TasdError {
    fn from(value: PacketError) -> Self {
        Self::Packet(value)
    }
}


#[derive(Debug, Clone, PartialEq)]
pub struct TasdFile {
    pub version: u16,
    pub keylen: u8,
    pub packets: Vec<Packet>,
    pub path: Option<PathBuf>,
}
impl Default for TasdFile {
    fn default() -> Self { Self {
        version: u16::from_be_bytes(LATEST_VERSION),
        keylen: 2,
        packets: vec![],
        path: None
    }}
}
impl TasdFile {
    pub fn new() -> Self {
        let mut tasd = Self::default();
        tasd.packets.push(
            DumpCreated {
                epoch: SystemTime::now().duration_since(UNIX_EPOCH).expect("Time has gone backwards?").as_secs() as i64
            }.into()
        );
        
        tasd
    }
    
    pub fn parse_file<P: Into<PathBuf>>(path: P) -> Result<Self, TasdError> {
        let path = path.into();
        let data = std::fs::read(&path)?;
        let mut file = Self::parse_slice(&data)?;
        file.path = Some(path);
        
        Ok(file)
    }
    
    pub fn parse_slice(data: &[u8]) -> Result<Self, TasdError> {
        let mut r = Reader::new(&data);
        if r.remaining() < 7 {
            return Err(TasdError::MissingHeader);
        }
        let magic = r.read_len(4);
        if magic != MAGIC_NUMBER {
            return Err(TasdError::MagicNumberMismatch(magic.to_vec()));
        }
        
        let mut file = Self {
            version: r.read_u16(),
            keylen: r.read_u8(),
            packets: vec![],
            path: None,
        };
        
        while r.remaining() > 0 {
            use PacketError::*;
            match Packet::with_reader(&mut r, file.keylen) {
                Ok(packet) => file.packets.push(packet),
                Err(err) => match err {
                    MissingKey | MismatchedKey | MissingPayloadLength | UnsupportedExponent(_) => return Err(err.into()),
                    InvalidPayload { key, payload } => println!("InvalidPayload! Skipping. ({key:02X?}, {payload:02X?}"),
                }
            }
        }
        
        Ok(file)
    }
    
    /// Encodes data in this [TasdFile] into a TASD formatted Vec of bytes.
    pub fn encode(&self) -> Vec<u8> {
        let mut w = Writer::new();
        
        w.write_slice(&MAGIC_NUMBER);
        w.write_slice(&LATEST_VERSION);
        w.write_u8(self.keylen);
        
        for packet in &self.packets {
            w.write_slice(&packet.encode(self.keylen));
        }
        
        w.to_vec()
    }
    
    /// Attempts to save this file to the path specified in [`self.path`][field@TasdFile::path].
    /// 
    /// If the path is `None`, or any IO errors are encountered, an `Err` is returned, otherwise `Ok(())`.
    pub fn save(&self) -> Result<(), TasdError> {
        if let Some(path) = self.path.as_ref() {
            std::fs::write(path, self.encode()).map_err(|err| err.into())
        } else {
            Err(TasdError::MissingPath)
        }
    }
}


src/lookup.rs

pub fn console_type_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "NES",
        0x02 => "SNES",
        0x03 => "N64",
        0x04 => "GC",
        0x05 => "GB",
        0x06 => "GBC",
        0x07 => "GBA",
        0x08 => "Genesis",
        0x09 => "A2600",
        0xFF => "Custom",
        _ => return None
    }.into())
}

pub fn console_region_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "NTSC",
        0x02 => "PAL",
        _ => return None
    }.into())
}

pub fn attribution_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "Author",
        0x02 => "Verifier",
        0x03 => "TASD File Creator",
        0x04 => "TASD File Editor",
        0xFF => "Other",
        _ => return None
    }.into())
}

pub fn memory_init_data_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "No initialization required",
        0x02 => "All 0x00",
        0x03 => "All 0xFF",
        0x04 => "00 00 00 00 FF FF FF FF (repeating)",
        0x05 => "Random",
        0xFF => "Custom",
        _ => return None
    }.into())
}

pub fn memory_init_device_lut(kind: u16) -> Option<String> {
    Some(match kind {
        0x0101 => "NES CPU RAM",
        0x0102 => "NES Cartridge Save Data",
        0x0201 => "SNES CPU RAM",
        0x0202 => "SNES Cartridge Save Data",
        0x0501 => "GB CPU RAM",
        0x0502 => "GB Cartridge Save Data",
        0x0601 => "GBC CPU RAM",
        0x0602 => "GBC Cartridge Save Data",
        0x0701 => "GBA CPU RAM",
        0x0702 => "GBA Cartridge Save Data",
        0x0801 => "Genesis CPU RAM",
        0x0802 => "Genesis Cartridge Save Data",
        0x0901 => "A2600 CPU RAM",
        0x0902 => "A2600 Cartridge Save Data",
        0xFFFF => "Custom/Other Device",
        _ => return None
    }.into())
}

pub fn game_identifier_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "MD5 Hash",
        0x02 => "SHA1 Hash",
        0x03 => "SHA224 Hash",
        0x04 => "SHA256 Hash",
        0x05 => "SHA384 Hash",
        0x06 => "SHA512 Hash",
        0x07 => "SHA512/224 Hash",
        0x08 => "SHA512/256 Hash",
        0x09 => "SHA3-224 Hash",
        0x0A => "SHA3-256 Hash",
        0x0B => "SHA3-384 Hash",
        0x0C => "SHA3-512 Hash",
        0x0D => "SHAKE-128 Hash",
        0x0E => "SHAKE-256 Hash",
        0xFF => "Other",
        _ => return None
    }.into())
}

pub fn identifier_encoding_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "Raw Binary",
        0x02 => "Base 16 (Case Insensitive)",
        0x03 => "Base 32 (Case Insensitive)",
        0x04 => "Base 64",
        _ => return None
    }.into())
}

pub fn controller_type_lut(kind: u16) -> Option<String> {
    Some(match kind {
        0x0101 => "NES Standard Controller",
        0x0102 => "NES Four Score",
        0x0103 => "(RESERVED) NES Zapper",
        0x0104 => "(RESERVED) NES Power Pad",
        0x0105 => "(RESERVED) Famicom Family BASIC Keyboard",
        0x0201 => "SNES Standard Controller",
        0x0202 => "SNES Super Multitap",
        0x0203 => "SNES Mouse",
        0x0204 => "(RESERVED) SNES Superscope",
        0x0301 => "N64 Standard Controller",
        0x0302 => "N64 Standard Controller with Rumble Pak",
        0x0303 => "N64 Standard Controller with Controller Pak",
        0x0304 => "N64 Standard Controller with Transfer Pak",
        0x0305 => "N64 Mouse",
        0x0306 => "(RESERVED) N64 Voice Recognition Unit (VRU)",
        0x0307 => "(RESERVED) N64 RandNet Keyboard",
        0x0308 => "N64 Densha de Go",
        0x0401 => "GC Standard Controller",
        0x0402 => "(RESERVED) GC Keyboard",
        0x0501 => "GB Gamepad",
        0x0601 => "GBC Gamepad",
        0x0701 => "GBA Gamepad",
        0x0801 => "Genesis (Mega Drive) 3-Button",
        0x0802 => "Genesis (Mega Drive) 6-Button",
        0x0901 => "A2600 Joystick",
        0x0902 => "(RESERVED) A2600 Paddle",
        0x0903 => "A2600 Keyboard Controller",
        0xFFFF => "Other/Unspecified",
        _ => return None
    }.into())
}

pub fn input_moment_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "Frame",
        0x02 => "Cycle Count",
        0x03 => "Milliseconds",
        0x04 => "Microseconds * 10",
        _ => return None
    }.into())
}

pub fn transition_index_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "Frame",
        0x02 => "Cycle Count",
        0x03 => "Milliseconds",
        0x04 => "Microseconds * 10",
        0x05 => "INPUT_CHUNK Index",
        _ => return None
    }.into())
}

pub fn transition_kind_lut(kind: u8) -> Option<String> {
    Some(match kind {
        0x01 => "Soft Reset",
        0x02 => "Power Reset",
        0x03 => "Restart TASD File",
        0xFF => "Packet Derived",
        _ => return None
    }.into())
}